cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/ScopeManager.kt

```diff
@@ -898,3 +976,170 @@ class ScopeManager : ScopeProvider {
+            return i
+        }
+
+    val isDirectMatch: Boolean
+        get() {
+            return this.casts?.all { it is DirectMatch } == true
+        }
+}
+
+data object IncompatibleSignature : SignatureResult()
+
+data class SignatureMatches(override val casts: List<CastResult>) : SignatureResult(casts)
+
+fun FunctionDeclaration.matchesSignature(
+    signature: List<Type>,
+    useDefaultArguments: Boolean = false,
+    call: CallExpression? = null,
+): SignatureResult {
+    val casts = mutableListOf<CastResult>()
+
+    var remainingArguments = signature.size
+
+    // Loop through all parameters of this function
+    for ((i, param) in this.parameters.withIndex()) {
+        // Once we are in variadic mode, all arguments match
+        if (param.isVariadic) {
+            remainingArguments = 0
+            break
+        }
+
+        // Try to find a matching call argument by index
+        val type =
+            signature.getOrNull(i)
+                ?: // If the argument is null, we could still have a default argument
+                return if (useDefaultArguments) {
+                    val defaultParam = this.defaultParameters[i]
+                    if (defaultParam != null) {
+                        casts += DirectMatch
+                        remainingArguments--
+                        continue
+                    } else {
+                        IncompatibleSignature
+                    }
+                } else {
+                    IncompatibleSignature
+                }
+
+        // Check, if we can cast the arg into our target type; and if, yes, what is the "distance"
+        // to the base type. We need this to narrow down the type during resolving
+        val match = type.tryCast(param.type, call?.arguments?.getOrNull(i), param)
+        if (match == CastNotPossible) {
+            return IncompatibleSignature
+        }
+
+        casts += match
+        remainingArguments--
+    }
+
+    // TODO(oxisto): In some languages, we can also have named parameters, but this is not yet
+    //  supported
+
+    return if (remainingArguments > 0) {
+        IncompatibleSignature
+    } else {
+        SignatureMatches(casts)
+    }
+}
+
+/**
+ * This is the result of [ScopeManager.resolveCall]. It holds all necessary intermediate results
+ * (such as [candidateFunctions], [viableFunctions]) as well as the final result (see [bestViable])
+ * of the call resolution.
+ */
+data class CallResolutionResult(
+    /** The original call expression. */
+    val call: CallExpression,
+
+    /**
+     * A set of candidate symbols we discovered based on the [CallExpression.callee] (using
+     * [SymbolResolver.findSymbols]), more specifically a list of [FunctionDeclaration] nodes.
+     */
+    var candidateFunctions: Set<FunctionDeclaration>,
+
+    /**
+     * A set of functions, that restrict the [candidateFunctions] to those whose signature match.
+     */
+    var viableFunctions: Set<FunctionDeclaration>,
```

Comment:
Maybe matchingFunctions would be a better name, signatureMatchingFunctions is a bit lengthy, but to me the difference between candidateFunctions and viable functions is not clear enough from the name alone.

Reply:
I took the naming from the C++ overload resolution algorithm: https://en.cppreference.com/w/cpp/language/overload_resolution#Viable_functions